텐서(Tensor)

> ## 학습 목표

- 파이토치 텐서의 기본 개념을 이해하고 텐서를 생성 및 활용할 수 있다.
- 텐서의 기본 연산 및 조작 방법을 익히고, 텐서의 속성과 특성에 대해 이해할 수 있다.
- 텐서 데이터 타입과 장치 이동 방식을 이해하고, 이를 활용하여 효율적인 프로그래밍을 할 수 있다.
- 텐서 변환과 브로드캐스팅 개념을 이해하고, 이를 활용하여 수치 계산 및 데이터 분석을 수행할 수 있다.

## 2.1 텐서(Tensor) 소개

- 최근의 머신 러닝 시스템은 일반적으로 **텐서를 기본 데이터 구조**로 사용한다.  
- 데이터를 위한 컨테이너라고 생각하면 쉽게 이해할 수 있다.
- NumPy는 훌륭한 프레임워크지만, GPU를 사용하여 수치 연산을 가속화할 수는 없기 때문에 GPU연산이 가능한 PyTorch에서의 Tensor를 사용한다. (디바이스 선택이 수월함.)

  심층 신경망에서 GPU는 종종 50배 또는 그 이상의 속도 향상을 제공하기 때문에, 안타깝게도 NumPy는 딥러닝 프로그래밍에는 충분치 않다.

- **PyTorch 텐서(Tensor)는 개념적으로 NumPy 배열과 동일** : 텐서(Tensor)는 n-차원 배열이며, PyTorch는 이러한 텐서들의 연산을 위한 다양한 기능들을 제공한다.


> ### 텐서의 특성

-   축의 개수:  
    랭크(Rank)라고도 부르며, 넘파이 배열에서는 `ndim`을 통해 확인할 수 있다.  
    
-   크기:  
    텐서의 각 축을 따라 얼마나 많은 차원이 있는지를 나타낸 튜플 넘파이 배열에서는 `shape`을 통해 확인할 수 있다.  

-   데이터 타입:  
    텐서에 포함된 데이터 타입이다.  
    타입은 float32, uint8, float64 등이 될 수 있으며 문자열은 지원하지 않는다. `dtype` 속성으로 데이터를 확인할 수 있다.

In [5]:
# 스칼라 (0차원 텐서)
# 하나의 숫자만을 담고 있는 텐서를 스칼라라고 하며 0차원 텐서라고 한다
# 스칼라의 축의 개수는? 0개

import torch
import numpy as np

scalar = np.array(10)
print(scalar)
print(scalar.ndim) # number of dimension >> 0
print(scalar.shape)
# 차원이 없으니까 빈 튜플 출력

10
0
()


In [7]:
# 벡터(vector) 1차원
# 숫자들의 배열을 벡터 (1차원 텐서)
# 벡터의 축의 개수? 1개

import torch
import numpy as np

vector = np.array([1,2,3,4,5])
print(vector)
print(vector.ndim)   # 차원 >> 1
print(vector.shape)  # 배열의 크기

[1 2 3 4 5]
1
(5,)


In [8]:
# 행렬(2D Tensor)
# 벡터들의 배열  >> 2차원 텐서
# 행렬에는 행, 열 2개의 축이 있음

import torch
import numpy as np

matrix = np.array([[1,2,3], [4,5,6]] )
print(matrix)
print(matrix.ndim)
print(matrix.shape) # (2, 3) 2행 3열

[[1 2 3]
 [4 5 6]]
2
(2, 3)


In [11]:
# 고차원 텐서
# 행렬들을 하나의 새로운 배열을 합치면 숫자로 채워진 직육면체가 되요. >> 3차원 텐서
# 딥러닝에서는 보통 5차원 텐서까지 다루게 됨

import torch
import numpy as np

tensor_3d = np.arange(24).reshape(2, 3, 4)

print(tensor_3d)        # 2개의 3*4 행렬로 구성
print(tensor_3d.ndim)   # 텐서의 차원 3
print(tensor_3d.shape)  # (2, 3, 4)

[[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]
3
(2, 3, 4)


> ### Tensor의 Shape

넘파이의 배열에서 `shape`를 이용하여 텐서의 크기를 알아볼 경우 return값은 다음과 같다  
  
**1차원 : (열,)**  
**2차원 : (행, 열)**  
**3차원 : (깊이, 행, 열)**  
  
즉, 열의 기준이 되어 차원이 늘어날수록 늘어난 차원의 크기가 열의 앞에 추가된다고 생각하면 된다.  
머신러닝에서 사용하는 train\_data의 shape은 다음과 같다.  

**1\. 벡터 데이터: (samples, features) 2D tensor**
집값 예측 문제라고 생각하고 주어진 데이터가 100개의 연식, 동네, 역세권의 유무에 따른 데이터라고 하면 (100, 3)크기의 텐서에 저장될 수 있다.  

**2\. 이미지: (samples, height, width, channels) 4D tensor**
채널 우선방식 과 채널 마지막 방식 으로 나뉘지만 보통의 경우 100 장의 28x28의 컬러 이미지라면 (100, 28, 28, 3)크기의 텐서에 저장될 수 있다.  

**3\. 동영상: (samples, Time(frames), height, width, channels) 5D tenesor**

- FPS (Frame per Second)

In [13]:
import numpy as np
from keras.datasets import mnist

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

print(train_images.shape)  # (60000, 28, 28) 28*28 이미지 60,000장
print(test_images.shape)   # (10000, 28, 28) 28*28 이미지 10,000장

(60000, 28, 28)
(10000, 28, 28)


In [17]:
use_images = train_images[:100, :, :]
print(use_images.shape)
# 일반적으로 딥러닝에서 사용하는 모든 데이터 텐서의 첫번째 축(axis0)은 샘플 축(sample axis)임
# >> 즉 샘플의 개수

(100, 28, 28)


위와 같은 샘플 축의 슬라이싱은 모델 수행 시 배치 데이터를 나눌 때 사용된다.

배치 데이터를 다룰 때는 첫번째 축(axis0)를 배치축(batch axis) 또는 배치차원(batch dimension)이라고 부른다.

In [18]:
# 6만장의 데이터 중에서 위쪽 절반만 사용하고 싶을 겨우
use_images = train_images[:, :14, :]
print(use_images.shape)

(60000, 14, 28)


#### ※ Tensor과 Ndarray
- 텐서 (Tensor)는 NumPy 어레이와 비슷하지만, 텐서는 GPU, TPU와 같은 가속기에서 사용할 수 있고, 텐서는 값을 변경할 수 없습니다.


먼저 Pytorch에서 주로 다루는 데이터 자료형인 Tensor과 Numpy에서 주로 다루는 데이터 자료형인 ndarray는 다차원 배열과 배열 생성 시 다양한 데이터 타입 적용이 가능하지만 주요한 차이점이 존재합니다.

우선 가장 중요한 차이점은 GPU를 활용한 연산이 가능한가? 이며, 그 외로 중요한 기능들에서 차이점이 있습니다.

- 텐서는 파이토치에서 스칼라, 벡터, 행렬 같은 개념으로 연산에 사용하는 기본적인 자료구조이며 다차원 배열로 표현할 수 있습니다.
- 파이썬 numpy의 ndarray와 유사합니다.(n차원의 배열 객체)
- 텐서는 n차원으로 구성되어 있는데 0차원은 스칼라, 1차원은 벡터, 2차원은 행렬, 3차원부터 텐서입니다.
- 1차원 텐서, 2차원 텐서, 3차원 텐서라고도 합니다.
- torch 패키지를 임포트하고 tensor 함수를 이용하여 텐서를 생성할 수 있습니다.

In [22]:
import torch

# 1차원 텐서 생성
a = torch.tensor([1,2,3,4,5])
print(a)
print(a.shape)
print(a.size()) # 텐서 크기 확인 size()
# size() 텐서의 차원, 각 차원의 크기를 알려줌

tensor([1, 2, 3, 4, 5])
torch.Size([5])
torch.Size([5])


In [26]:
import torch

# 2차원 텐서 생성
# 1차원 텐서들을 바깥 대괄호([]) 감싸기

x1 = torch.tensor([ [1,2,3],
                    [4,5,6],
                    [7,8,9] ])

print(x1)
print(x1.size())
# torch.Size([3, 3]) 3개의 행과 3개의 열로 이루어진 3*3 행렬 형태의 텐서
print(x1.ndim)

tensor([[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]])
torch.Size([3, 3])
2


In [27]:
import torch

# 3행 2열 리스트로 2차원 텐서 생성하기

data = [[1, 2],[3, 4],[5, 6]]
# 차원 리스트를 만듭니다. 이 리스트는 3개의 내부 리스트로 구성되어 있으며,
# 각 내부 리스트는 2개의 숫자를 포함하고 있습니다.

# 텐서로 변환
x2 = torch.tensor(data)
#  data 리스트를 파이토치 텐서로 변환합니다.
# torch.tensor() 함수는 파이썬 리스트를 받아 텐서로 변환해줍니다.

print(x2)

print(x2.shape)

tensor([[1, 2],
        [3, 4],
        [5, 6]])
torch.Size([3, 2])


In [32]:
import torch

# 3차원 텐서 생성
# 2차원 텐서를 대괄호 감싸기

x3 = torch.tensor([[[1,2,3], [4,5,6], [7,8,9]]])
print(x3)

# 3차원 텐서 >> 큐브 cube, 박스 box 구조 가짐
# 가장 바깥쪽 대괄호가 첫번째 차원 의미(깊이) : 1
# 중간 대괄호는 두번째 차원 (행) : 3
# 가장 안쪽 대괄호는 세번째 차원 (열) : 3

print(x3.size())

tensor([[[1, 2, 3],
         [4, 5, 6],
         [7, 8, 9]]])
torch.Size([1, 3, 3])


In [36]:
# Numpy 배열과 PyTorch 텐서 간 변환

import torch
import numpy as np

x3_1 = [[1,2,3], [4,5,6], [7,8,9]]
data_np = np.array(x3_1)

# torch.from_numpy() : numpy >> torch
torch.from_numpy(data_np)

tensor([[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]])

특정 값으로 텐서 생성하기

In [37]:
import torch

# 0으로 채워진 텐서
torch.zeros(2,3)

tensor([[0., 0., 0.],
        [0., 0., 0.]])

In [38]:
# 1로 채워진 텐서
torch.ones(2,3)

tensor([[1., 1., 1.],
        [1., 1., 1.]])

In [39]:
# 특정 값으로 채워진 텐서
torch.full((2,3), 7)
# 2행 3열의 2차원 텐서 생성 >> 그 모든 원소를 7로 채워줘

tensor([[7, 7, 7],
        [7, 7, 7]])

### `Tip!`
- 이러한 함수들은 특정 크기와 초기값을 가진 텐서를 빠르게 생성할 때 유용합니다.
- 예를 들어, 신경망의 가중치를 초기화하거나 마스크를 만들 때 자주 사용됩니다. 각 함수의 특징을 정리하면 다음과 같습니다:
  - torch.zeros(): 모든 원소가 0인 텐서 생성
  - torch.ones(): 모든 원소가 1인 텐서 생성
  - torch.full(): 모든 원소가 지정한 값으로 채워진 텐서 생성
- 이러한 함수들은 텐서의 크기를 인자로 받으며, torch.full()의 경우 추가로 채울 값을 인자로 받습니다.

In [43]:
# random 한 텐서 생성

import torch

random_tensor = torch.randn((2,3))
# 평균이 0, 표준편차 1 인 표준 정규분포에서 무작위로 숫자 추출
# (2,3) : 생성될 텐서 형태 지정

print(random_tensor)

tensor([[ 1.0978, -0.0993, -0.2362],
        [ 0.5607,  0.7255,  0.4813]])


In [44]:
import torch

torch.randn_like(random_tensor)
# torch.randn_like
# 주어진 텐서와 동일한 크기, 데이터 타입 가진 새로운 텐서 생성

tensor([[ 0.6703,  1.0845, -1.6237],
        [ 1.2045, -0.3629,  0.4592]])

### `Tip!`

주요 특징
- 크기 유지: tensor_like는 random_tensor와 정확히 같은 크기(2x3)를 가집니다.
- 독립적인 랜덤 값: 생성된 값들은 random_tensor의 값들과 완전히 독립적입니다.
- 데이터 타입 일치: tensor_like는 random_tensor와 동일한 데이터 타입을 가집니다.
- GPU 호환성: 만약 random_tensor가 GPU에 있다면, tensor_like도 자동으로 GPU에 생성됩니다.

※ 이러한 기능은 딥러닝 모델에서 가중치 초기화나 노이즈 생성 등 다양한 상황에서 유용하게 사용될 수 있습니다.

In [47]:
import torch

tensor_rand = torch.rand(2,3)
# rand(행, 열) 0-1 사이 실수값으로 채워진 텐서

print(tensor_rand)

tensor([[0.7435, 0.4562, 0.0939],
        [0.3889, 0.7349, 0.4837]])


In [50]:
torch.arange(1,7).reshape(2,3)

tensor([[1, 2, 3],
        [4, 5, 6]])

### `Tip!`
- rand()는 무작위 값으로 텐서를 채우는 데 유용하며, arange()와 reshape()의 조합은 특정 패턴이나 순서를 가진 텐서를 만드는 데 사용됩니다.

텐서 속성

- 텐서의 차원을 확인하는 dim() 메서드가 있습니다. 이는 텐서의 rank라고도 불립니다.
- 텐서의 형상(shape)을 확인하는 shape 속성이 있습니다. 이는 각 차원의 크기를 튜플 형태로 반환합니다.
- 텐서의 크기를 확인하는 size() 메서드도 있습니다. 이는 shape와 동일한 정보를 반환합니다.
- 데이터타입 자료형을 확인하는 dtype 속성은 datatype() 속성이며, 이를 통해 텐서의 데이터 타입(예: 정수, 실수, 불리언 등)을 알 수 있습니다.
- 텐서가 저장된 장치(CPU 또는 GPU)를 확인하는 device 속성이 있습니다.
- 텐서의 총 원소 개수를 반환하는 numel() 메서드도 유용합니다.

In [56]:
import torch

x = torch.randn(3,4,5)

print(x.dim()) # 3 >> 텐서의 차원
print(x.shape)
print(x.size())
print(x.dtype) # torch.float32 (텐서의 기본 데이터 타입)
print(x.device)
print(x.numel())
# numel() : number of elements 텐서의 총 원소 개수 3*4*5 = 60

3
torch.Size([3, 4, 5])
torch.Size([3, 4, 5])
torch.float32
cpu
60


In [61]:
import torch

# 1차원 텐서
# 텐서의 크기 확인 size(), shape
# 차원 확인 dim(), ndimension()
# 데이터 타입 dtype()

a = torch.tensor([1,2,3,4,5])
print(a)
print(a.size())
print(a.shape)
print(a.dim())
print(a.ndimension()) # 차원(rank)
print(a.dtype)

tensor([1, 2, 3, 4, 5])
torch.Size([5])
torch.Size([5])
1
1
torch.int64


In [64]:
import torch

# 실수형 FloatTensor

x4 = torch.FloatTensor([[1,2],[3,4]])

print(x4)
print(x4.size())
print(x4.shape)
print(x4.dim())
print(x4.ndimension())
print(x4.dtype)

tensor([[1., 2.],
        [3., 4.]])
torch.Size([2, 2])
torch.Size([2, 2])
2
2
torch.float32


In [66]:
import torch

# 3차원 텐서
x5 = torch.tensor([[[1,2,3],[4,5,6],[7,8,9]]])

print(x5)
print(x5.size())
print(x5.shape)
print(x5.dim())
print(x5.ndimension())
print(x5.dtype)

tensor([[[1, 2, 3],
         [4, 5, 6],
         [7, 8, 9]]])
torch.Size([1, 3, 3])
torch.Size([1, 3, 3])
3
3
torch.int64


텐서 생성 실습

- 이 예제는 행렬 곱셈의 기본 원리를 보여줍니다.
- 첫 번째 행렬의 열 수(3)와 두 번째 행렬의 행 수(3)가 일치해야 행렬 곱셈이 가능합니다.
- 결과 텐서의 크기는 첫 번째 행렬의 행 수(2)와 두 번째 행렬의 열 수(2)로 결정됩니다.

In [71]:
import torch

# [2,3]과 [3,2] 텐서 생성
a = torch.tensor([[1,2,3], [4,5,6]])
b = torch.tensor([[7,8],[9,10],[11,12]])

c = torch.matmul(a,b)
print(c)
print(c.shape)

tensor([[ 58,  64],
        [139, 154]])
torch.Size([2, 2])


- 이 코드는 두 텐서 a와 b의 행렬 곱셈을 수행합니다. 행렬 곱셈 과정을 자세히 설명하겠습니다:
- 텐서 a는 2x3 크기이고, 텐서 b는 3x2 크기입니다.
- 행렬 곱셈 규칙에 따라, 첫 번째 행렬의 열 수(3)와 두 번째 행렬의 행 수(3)가 일치해야 합니다.
- 결과 텐서 c의 크기는 첫 번째 행렬의 행 수(2)와 두 번째 행렬의 열 수(2)가 됩니다. 따라서 c는 2x2 크기가 됩니다.
- 행렬 곱셈 과정:
  - c[0,0] = a[0,0]b[0,0] + a[0,1]b[1,0] + a[0,2]b[2,0]
    = 17 + 29 + 311 = 7 + 18 + 33 = 58
  - c[0,1] = a[0,0]b[0,1] + a[0,1]b[1,1] + a[0,2]b[2,1]
    = 18 + 210 + 312 = 8 + 20 + 36 = 64
  - c[1,0] = a[1,0]b[0,0] + a[1,1]b[1,0] + a[1,2]b[2,0]
    = 47 + 59 + 611 = 28 + 45 + 66 = 139
  - c[1,1] = a[1,0]b[0,1] + a[1,1]b[1,1] + a[1,2]b[2,1]
    = 48 + 510 + 612 = 32 + 50 + 72 = 154

In [78]:
import torch

tensor1 = torch.tensor([1,2,3])
tensor2 = torch.zeros(3,3)
tensor3 = torch.ones(2,2)
tensor4 = torch.rand(3,3)

reshaped = tensor4.reshape(1, 9)
print(reshaped)
transposed = tensor4.t()
print(transposed)

tensor([[0.7626, 0.2694, 0.0393, 0.1890, 0.2490, 0.9179, 0.8921, 0.7788, 0.7716]])
tensor([[0.7626, 0.1890, 0.8921],
        [0.2694, 0.2490, 0.7788],
        [0.0393, 0.9179, 0.7716]])


In [82]:
import torch

# 실수형 텐서
float_tensor = torch.tensor([1.5, 2.5, 3.5], dtype=torch.float32)
print(float_tensor)

# 정수형 텐서
int_tensor = torch.tensor([1,2,3], dtype=torch.int64)
print(int_tensor)

tensor([1.5000, 2.5000, 3.5000])
tensor([1, 2, 3])


#### 기본 연산

-   **덧셈과 뺄셈**: 같은 크기의 텐서 간에 요소별로 수행됩니다.
-   **요소별 곱셈**: 두 텐서의 대응하는 요소끼리 곱합니다.
-   **스칼라 곱**: 텐서의 모든 요소에 스칼라 값을 곱합니다.

#### 행렬 곱셈

-   행렬 곱셈은 `torch.matmul()` 또는 `@` 연산자를 사용합니다.
-   왼쪽 행렬의 열 수와 오른쪽 행렬의 행 수가 일치해야 합니다.
-   결과 행렬의 크기는 (왼쪽 행렬의 행 수) x (오른쪽 행렬의 열 수)가 됩니다.

#### 고급 연산

-   **전치(Transpose)**: 행과 열을 바꾸는 연산으로, `t()` 메서드를 사용합니다.
-   **차원 축소**: `sum()`, `mean()` 등의 함수로 특정 차원을 따라 값을 집계합니다.
-   **브로드캐스팅**: 크기가 다른 텐서 간의 연산을 자동으로 조정합니다.

#### 비선형 활성화 함수

-   ReLU, Sigmoid, Tanh 등의 함수를 텐서에 적용하여 비선형성을 도입합니다.

이러한 연산들은 신경망의 순전파와 역전파 과정에서 핵심적인 역할을 합니다.

효율적인 텐서 연산은 딥러닝 모델의 성능과 학습 속도에 직접적인 영향을 미칩니다.

In [84]:
import torch
import numpy as np

# 1차 텐서 덧셈
a = torch.tensor([1,2,3])
b = torch.tensor([4,5,6])

c = a + b
print(c)
print(c.shape)

tensor([5, 7, 9])
torch.Size([3])


In [85]:
# 1차 텐서 뺄셈
a = torch.tensor([1,2,3])
b = torch.tensor([4,5,6])

c = a - b
print(c)
print(c.shape)

tensor([-3, -3, -3])
torch.Size([3])


In [89]:
a = torch.tensor([1,2,3])
b = torch.tensor([4,5,6])

# 기본 연산
addition = a + b
multiplication = a * b
dot_product = torch.dot(a,b)

print(addition)
print(multiplication)
print(dot_product)

tensor([5, 7, 9])
tensor([ 4, 10, 18])
tensor(32)


In [91]:
import torch

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
tensor = torch.rand(3,3).to(device)
print(f'Tensor is on {device}')

Tensor is on cpu


이 코드는 PyTorch를 사용하여 텐서를 생성하고 GPU가 사용 가능한 경우 GPU로 텐서를 이동시키는 예제입니다. 각 줄을 자세히 설명하겠습니다:

1.  `device = torch.device("cuda" if torch.cuda.is_available() else "cpu")`  
    이 줄은 GPU 사용 가능 여부를 확인하고 적절한 장치를 선택합니다:
    
    -   `torch.cuda.is_available()`은 CUDA(GPU 연산을 위한 NVIDIA의 병렬 컴퓨팅 플랫폼)가 사용 가능한지 확인합니다.
    -   GPU가 사용 가능하면 `device`는 "cuda"로 설정되고, 그렇지 않으면 "cpu"로 설정됩니다.
    
2.  `tensor = torch.rand(3, 3).to(device)`  
    이 줄은 3x3 크기의 랜덤 텐서를 생성하고 선택된 장치로 이동시킵니다:
    
    -   `torch.rand(3, 3)`는 0과 1 사이의 균일 분포에서 무작위로 값을 추출하여 3x3 텐서를 생성합니다.
    -   `.to(device)`는 생성된 텐서를 선택된 장치(GPU 또는 CPU)로 이동시킵니다.
    
3.  `print(f"Tensor is on {device}")`  
    이 줄은 텐서가 어떤 장치에 있는지 출력합니다.

이 코드는 GPU가 사용 가능한 경우 자동으로 GPU를 활용하여 연산을 수행할 수 있게 해줍니다. GPU를 사용하면 대규모 텐서 연산과 딥러닝 모델 학습 속도를 크게 향상시킬 수 있습니다. GPU가 없는 시스템에서는 자동으로 CPU를 사용하므로, 코드의 호환성도 유지됩니다.

In [95]:
import torch
import numpy as np

a = torch.tensor([10,20,30])
b = torch.tensor([2,4,5])

result = a/b
print(result)

tensor([5., 5., 6.])


In [97]:
a = torch.tensor([2,4,5])
result = a ** 2 # 거듭 제곱
print(result)

tensor([ 4, 16, 25])


In [102]:
# 집계함수

tensor = torch.tensor([[1,2,3], [4,5,6]])
print(tensor)
# 전체 합계
sum_value = torch.sum(tensor)
print(sum_value)

tensor([[1, 2, 3],
        [4, 5, 6]])
tensor(21)


In [101]:
# 특정 차원 합계
torch.sum(tensor, dim=0) # 행 기준으로 열의 합

tensor([5, 7, 9])

In [103]:
torch.sum(tensor, axis=0)

tensor([5, 7, 9])

In [104]:
torch.sum(tensor, 0)

tensor([5, 7, 9])

In [108]:
import torch
import numpy as np

tensor = torch.tensor([1,2,3,4,5])

# 최대값
max_value = torch.max(tensor) # tensor(5)
print(max_value)

# 최소값
min_value = torch.min(tensor) # tensor(1)
print(min_value)

tensor(5)
tensor(1)


In [110]:
# 비교 연산
condition = (tensor > 3)
print(condition)

tensor([False, False, False,  True,  True])


텐서 연산 위한 텐서 변환

- 텐서 연산은 필요에 따라 차원(dimensions)을 변경할 수 있습니다.
- view(), unsqueeze(), squeeze() 함수를 이용하여 텐서 변환을 할 수 있습니다.
- 중요한건 차원크기가 무조건 1로 늘어나거나 1인 것만 줄일 수 있습니다.

In [119]:
import torch

exam = torch.rand(3,1,4,1)
print(f'exam 텐서의 차원: {exam.shape}')

# 3, 1, 4, 1 차원 idx 0,1,2,3
exam = exam.squeeze(dim=1) # 차원 축소
print(f'exam 텐서의 차원: {exam.shape}')

# idx 는 리스트의 인덱싱처럼 뒤에서도 순번 지정 가능
exam = exam.squeeze(dim=-1)
print(f'exam 텐서의 차원: {exam.shape}')
# print(exam)

# 차원 늘리는 것(차원 확장) 인자값에 해당하는 순번에 '1'로 빈 차원이 삽입
exam = exam.unsqueeze(dim=0)
print(f'exam 텐서의 차원: {exam.shape}')

exam 텐서의 차원: torch.Size([3, 1, 4, 1])
exam 텐서의 차원: torch.Size([3, 4, 1])
exam 텐서의 차원: torch.Size([3, 4])
exam 텐서의 차원: torch.Size([1, 3, 4])



`Tensor`자료형의 '확장'은 무조건 차원 확장 후 → 차원값 확장 으로 진행되기 때문이다

> `torch.repeat()`, `torch.expand()`

위에서 차원 확장&축소와 관련된 메서드 `squeeze()` & `unsqueeze()`를 설명했으니 이제 확장된 차원크기를 조정하는 메서드를 알아보자

In [126]:
origin_data = torch.randint(0, 4, (3,2))
print(origin_data)
print()
# 차원 확장 [3,2] >> [1,3,2]
temp_data = origin_data.unsqueeze(0)
print(temp_data)
print()
# 확장된 차원을 반복해서 늘리고 싶어요 [1,3,2] >> [4,3,2]
exp_res_data = temp_data.expand(4,3,2)
print(exp_res_data)
print()
# 차원 사이즈가 어떻게 늘어났는지 확인
print(exp_res_data.shape)


tensor([[0, 0],
        [3, 1],
        [3, 0]])

tensor([[[0, 0],
         [3, 1],
         [3, 0]]])

tensor([[[0, 0],
         [3, 1],
         [3, 0]],

        [[0, 0],
         [3, 1],
         [3, 0]],

        [[0, 0],
         [3, 1],
         [3, 0]],

        [[0, 0],
         [3, 1],
         [3, 0]]])

torch.Size([4, 3, 2])


In [128]:
# repeat()
temp_data.shape

torch.Size([1, 3, 2])

In [131]:
rep_res_data = temp_data.repeat(4,1,1)
print(rep_res_data)
# 확장된 차원을 repeat 메서드로 늘림
# temp_data.repeat(4,1,1)
# dim = 0 : 4로 설정 : 첫번째 차원 4번 복제
# dim =1, dim=2 1로 설정 : 그대로 유지

print(rep_res_data.shape)

tensor([[[0, 0],
         [3, 1],
         [3, 0]],

        [[0, 0],
         [3, 1],
         [3, 0]],

        [[0, 0],
         [3, 1],
         [3, 0]],

        [[0, 0],
         [3, 1],
         [3, 0]]])
torch.Size([4, 3, 2])


`expand`는 차원 사이즈가 늘어나도 메모리는 공유되고 있기에  
하나의 원소값을 변조하면 공유 메모리로 묶여있는 다른 원소도 같이 영향받는다.

따라서 `expand`로 차원을 늘리는 경우는 브로드 캐스팅 방식으로 읽기전용 데이터를 만들 때 최적화 측면에서 사용하는 메서드라 보면 된다.

반대로 `repeat`메서드는 차원사이즈가 늘어나면 복제된 데이터는 완전히 독립적으로 운용되어 원소 변조가 발생하는 쓰기데이터일 때 위 메서드를 사용한다.

> ### 주요 텐서 변환 함수

#### view() 함수

-   텐서의 크기를 변경하는 데 사용됩니다.
-   변경 후에도 텐서 원소의 총 개수는 동일해야 합니다.
-   예: `x1_tensor = x_tensor.view(2, 3)`는 3x2 텐서를 2x3 텐서로 변환합니다.

#### unsqueeze() 함수

-   지정된 위치에 새로운 차원을 추가합니다.
-   `dim` 매개변수로 차원을 추가할 위치를 지정합니다.
-   예: `uns_tensor = x_tensor.unsqueeze(0)`는 첫 번째 위치에 새 차원을 추가합니다.

#### squeeze() 함수

-   크기가 1인 차원을 제거합니다.
-   특정 연산 후 불필요한 차원을 제거할 때 유용합니다.
-   예: `s_tensor = uns_tensor.squeeze()`는 크기가 1인 모든 차원을 제거합니다.

> ### 추가적인 텐서 변환 기능

#### chunk() 함수

-   텐서를 지정된 수의 덩어리로 나눕니다.
-   예: `chunks = torch.chunk(tensor, 3)`는 텐서를 3개의 덩어리로 나눕니다.

#### split() 함수

-   텐서를 지정된 크기의 덩어리로 나눕니다.
-   예: `splits = torch.split(tensor, 2)`는 텐서를 2개의 원소를 가진 덩어리들로 나눕니다.

이러한 텐서 변환 함수들은 딥러닝 모델 구축 및 데이터 전처리 과정에서 매우 유용하게 사용됩니다.

특히 복잡한 신경망 구조에서 텐서의 형태를 조정하거나, 배치 처리를 위해 데이터의 차원을 조정할 때 자주 활용됩니다.

In [136]:
# view() 함수: 텐서 크기 변경시 사용
# 변경 후에도 텐서 원소 개수는 반드시 같아야 함
import torch
import numpy as np

# randn() 함수 사용, 정규분포 텐서 생성
x_tensor = torch.randn(3,2)
# 평균 0, 표준편차 1 정규분포 >> 무작위 추출
print(x_tensor)
print(x_tensor.shape)
print()

# 텐서 크기 변경
x1_tensor = x_tensor.view(2,3)
print(x1_tensor)
print(x1_tensor.shape)

tensor([[-1.7911,  1.3056],
        [ 0.2732,  1.6177],
        [ 0.8047, -0.3105]])
torch.Size([3, 2])

tensor([[-1.7911,  1.3056,  0.2732],
        [ 1.6177,  0.8047, -0.3105]])
torch.Size([2, 3])


In [139]:
# 2차원 텐서
x_tensor = torch.randn(3,2)
print(x_tensor)
print()
# 1차원으로 텐서 차원 변경
x_tensor.view(-1)

tensor([[ 3.2195, -0.2955],
        [ 0.7263, -0.1090],
        [ 0.6259,  0.9416]])



tensor([ 3.2195, -0.2955,  0.7263, -0.1090,  0.6259,  0.9416])

In [142]:
# chunk / split

import torch
import numpy as np

# 1*6 tensor >> 3개의 텐서로 나누기
tensor = torch.tensor([1,2,3,4,5,6])
print(tensor)

# 텐서 분할
chunks = torch.chunk(tensor, 3)

for chunk in chunks:
    print(chunk)

tensor([1, 2, 3, 4, 5, 6])
tensor([1, 2])
tensor([3, 4])
tensor([5, 6])


`torch.chunk()` 함수의 특징:

-   가능한 한 균등하게 텐서를 나누려고 시도합니다.
-   텐서의 크기가 chunk 수로 나누어 떨어지지 않을 경우, 마지막 chunk가 더 작을 수 있습니다.
-   각 chunk는 원본 텐서의 뷰(view)이므로, 메모리를 추가로 사용하지 않습니다.

이 기능은 대규모 텐서를 처리할 때나 병렬 처리를 위해 데이터를 나눌 때 유용하게 사용될 수 있습니다.

In [145]:
import torch
import numpy as np

splits = torch.split(tensor,2) # 각 텐서가 2개의 원소 가짐
print(splits)
# torch.split() : 텐서를 지정된 크기의 덩어리로 나눔

for split in splits:
    print(split)

(tensor([1, 2]), tensor([3, 4]), tensor([5, 6]))
tensor([1, 2])
tensor([3, 4])
tensor([5, 6])


`torch.split()` 함수의 특징:

-   지정된 크기(여기서는 2)만큼 텐서를 균등하게 나눕니다.
-   마지막 덩어리가 지정된 크기보다 작을 수 있습니다(이 예제에서는 모든 덩어리가 동일한 크기입니다).
-   각 분할된 텐서는 원본 텐서의 뷰(view)이므로, 메모리를 추가로 사용하지 않습니다.

이 기능은 대규모 텐서를 처리할 때나 배치 처리를 위해 데이터를 나눌 때 유용하게 사용될 수 있습니다.

차원 추가

In [150]:
# unsqueeze() 차원 추가
# unsqueeze(x_tensor, dim)
# dim=0 >> 텐서 크기에서 인덱스 0, 즉 첫번째 자리에 차원을 추가하라는 의미예요

import torch
import numpy as np

x_tensor = torch.randn(3,2)

uns_tensor = x_tensor.unsqueeze(0)
# 첫번째 위치 (가장 바깥쪽)에 새로운 차원 추가

print(uns_tensor)
print(uns_tensor.shape)
print()
# dim=1 차원을 추가할 두 번째 자리
uns1_tensor = x_tensor.unsqueeze(1)
print(uns1_tensor)
print(uns1_tensor.shape)

tensor([[[ 1.4849, -0.1952],
         [-0.9684,  1.6734],
         [-1.3297,  0.7301]]])
torch.Size([1, 3, 2])

tensor([[[ 1.4849, -0.1952]],

        [[-0.9684,  1.6734]],

        [[-1.3297,  0.7301]]])
torch.Size([3, 1, 2])


- squeeze() 함수는 텐서에서 크기가 1인 차원을 제거하는 기능을 합니다.
- 특정 연산 후 불필요한 차원을 제거할 때 사용됩니다.
- 기본적으로 크기가 1인 차원 모두 찾아내서 없애 버려요

In [159]:
import torch
import numpy as np

# 차원 축소
s_tensor = uns_tensor.squeeze()
print(uns_tensor)
print(uns_tensor.shape)
print(s_tensor)
print(s_tensor.shape)

print()
print(uns1_tensor)
s1_tensor = uns1_tensor.squeeze()
print(uns1_tensor.shape)
print(s1_tensor)
print(s1_tensor.shape)

tensor([[[ 1.4849, -0.1952],
         [-0.9684,  1.6734],
         [-1.3297,  0.7301]]])
torch.Size([1, 3, 2])
tensor([[ 1.4849, -0.1952],
        [-0.9684,  1.6734],
        [-1.3297,  0.7301]])
torch.Size([3, 2])

tensor([[[ 1.4849, -0.1952]],

        [[-0.9684,  1.6734]],

        [[-1.3297,  0.7301]]])
torch.Size([3, 1, 2])
tensor([[ 1.4849, -0.1952],
        [-0.9684,  1.6734],
        [-1.3297,  0.7301]])
torch.Size([3, 2])


In [164]:
import torch

# 크기 텐서
x = torch.zeros(1,3,1)
print(x)
print(x.squeeze())

tensor([[[0.],
         [0.],
         [0.]]])
tensor([0., 0., 0.])
torch.Size([3])


In [166]:
print(x)
print(x.squeeze(dim=0))
print(x.squeeze(dim=0).shape)

tensor([[[0.],
         [0.],
         [0.]]])
tensor([[0.],
        [0.],
        [0.]])
torch.Size([3, 1])


브로드캐스팅

- 브로드캐스팅(broadcasting)은 두 텐서 간의 연산이 발생할 때, 텐서의 크기가 일치하지 않으면 연산이 가능하도록 자동으로 크기를 맞춰줍니다.
- 차원 크기가 1인 텐서는 그 차원의 크기를 다른 텐서의 크기에 맞춰 확장됩니다.
- 브로드캐스팅은 실제로 텐서를 확장하지 않고 연산만을 처리하기 때문에 메모리 낭비를 줄일 수 있습니다.

In [168]:
# 두 텐서 간 차원의 크기가 같거나 1차원
# b 의 텐서 크기가 a의 텐서 크기에 자동으로 맞추어줌

a = torch.tensor([[1,2],[3,5]])
b = torch.tensor([1,2])

c = a + b  # (2,2) + (2,) >> (2,2) 브로드캐스팅

print(a.shape)
print(b.shape)
print(c)

torch.Size([2, 2])
torch.Size([2])
tensor([[2, 4],
        [4, 7]])


In [169]:
c = a * b

print(a.shape)
print(b.shape)
print(c)

torch.Size([2, 2])
torch.Size([2])
tensor([[ 1,  4],
        [ 3, 10]])


In [175]:
# 3차원 텐서와 2차원 텐서
a = torch.randn(3,4,5) # 크기 (3, 4행, 5열)
b = torch.randn(4,5)   # 크기 (4,5)

c = a + b  #(3,4,5) + (4,5) >> (3,4,5) 브로드캐스팅
print(a)
print(b)
print(c.shape)
# 즉, b (4,5)가 (1,4,5)로 자동 확장

tensor([[[ 0.8201,  1.6324, -3.0609, -1.0946,  0.9656],
         [-0.4064, -1.7930,  1.1391, -1.5974, -2.0214],
         [-1.4189,  0.0939, -0.5246,  0.8328,  1.0083],
         [ 0.1953, -1.0371, -1.9896, -0.5132,  1.5840]],

        [[ 1.4416, -0.7297, -0.8988, -0.0948,  2.2640],
         [ 0.5780, -0.1246,  0.5309,  0.2679,  0.0291],
         [ 0.2086,  0.4903,  0.4187,  0.1624,  0.1039],
         [ 0.5231, -0.7162,  0.2146,  1.6841,  0.0386]],

        [[-0.4273,  0.7685, -0.9335, -0.9277,  0.7945],
         [-0.1485, -0.0720, -0.5347, -0.2300,  0.2617],
         [-1.4581, -0.4602,  1.5017, -0.1026,  1.0532],
         [ 1.3792, -0.2612,  2.4117,  0.8717,  0.4459]]])
tensor([[-0.6711,  1.3077,  0.0217,  0.5724, -0.8940],
        [ 0.0752, -0.5180,  0.4684,  0.1009, -0.7153],
        [ 0.9351, -0.9564, -0.8985, -0.3556,  0.1441],
        [-0.3628, -1.2015, -1.5491, -0.1148, -0.9336]])
torch.Size([3, 4, 5])


In [180]:
import torch

tensor = torch.tensor([[1,2,3],[4,5,6]])

# 첫번째 행만 선택
slice1 = tensor[0, :]
print(slice1)

# 두번째 열만 선택
slice2 =tensor[:, 1]
print(slice2)

tensor([1, 2, 3])
tensor([2, 5])


In [182]:
# 두번째 행과 세번째 열 원소 선택
slice3 = tensor[1,2]
print(slice3)

tensor(6)


In [184]:
# 첫번째 행의 첫번째와 두번째 열을 선택
slice_4 = tensor[0, :2]
print(slice_4)

tensor([1, 2])


In [186]:
import torch

# 2*2 tensor 두 개 생성
a = torch.tensor([[1,2],[3,4]])
b = torch.tensor([[5,6],[7,8]])

# 행기준(0번 축)으로 연결
torch.cat((a, b), dim=0)

tensor([[1, 2],
        [3, 4],
        [5, 6],
        [7, 8]])

In [187]:
# 열기준(1번 축)으로 연결
torch.cat((a, b), dim=1)

tensor([[1, 2, 5, 6],
        [3, 4, 7, 8]])

In [188]:
# eos